Support auto-configuration of the Datastore emulator (#1661)
spring-cloud/spring-cloud-gcp: docs/src/main/asciidoc/datastore.adoc

| `spring.cloud.gcp.datastore.credentials.encoded-key` | Base64-encoded OAuth2 credentials for authenticating with the Google Cloud Datastore API, if different from the ones in the <<spring-cloud-gcp-core,Spring Cloud GCP Core Module>> | No |
| `spring.cloud.gcp.datastore.credentials.scopes` | https://developers.google.com/identity/protocols/googlescopes[OAuth2 scope] for Spring Cloud GCP Cloud Datastore credentials | No | https://www.googleapis.com/auth/datastore
| `spring.cloud.gcp.datastore.namespace` | The Cloud Datastore namespace to use | No | the Default namespace of Cloud Datastore in your GCP project

> | `spring.cloud.gcp.datastore.host` | The `hostname:port` of the datastore service to connect. Can be used to connect to a https://cloud.google.com/datastore/docs/tools/datastore-emulator[Datastore Emulator]. If emulator are enabled and this property are empty, it will used as `localhost:<emulator_port>` | No |

Reviewer: I don't think this setting should be here. We're not advocating using any Cloud Datastore host other than the real one the client library uses by default. The only other host I think is reasonable to specify is the emulator, but you already have 2 settings for that below.

Reviewer: @meltsufin WDYT

Author: I agree with that. Do you guys think people will use the emulator on a different host than the client using the emulator itself?

Author: I agree with that. Do you think people will use the emulator on a different host? If not, we can just have the port and always connect to the localhost if the auto emulator is enabled.

If there is a chance to people trying to connect to external emulators, then we need to think abouut this property.

> | `spring.cloud.gcp.datastore.host` | The `hostname:port` of the datastore service to connect. Can be used to connect to a https://cloud.google.com/datastore/docs/tools/datastore-emulator[Datastore Emulator]. If the emulator is enabled and this property is not set, it will use `localhost:<emulator_port>` | No |

Reviewer 2: ```suggestion
| `spring.cloud.gcp.datastore.host` | The `hostname:port` of the datastore service or emulator to connect to. Can be used to connect to a manually started https://cloud.google.com/datastore/docs/tools/datastore-emulator[Datastore Emulator]. If the autoconfigured emulator is enabled, this property will be ignored and `localhost:<emulator_port>` will be used. | No |
```

Author: OK

| `spring.cloud.gcp.datastore.credentials.scopes` | https://developers.google.com/identity/protocols/googlescopes[OAuth2 scope] for Spring Cloud GCP Cloud Datastore credentials | No | https://www.googleapis.com/auth/datastore
| `spring.cloud.gcp.datastore.namespace` | The Cloud Datastore namespace to use | No | the Default namespace of Cloud Datastore in your GCP project
| `spring.cloud.gcp.datastore.emulator-host` | The `hostname:port` of the https://cloud.google.com/datastore/docs/tools/datastore-emulator[Datastore Emulator] to connect to | No |

> | `spring.cloud.gcp.datastore.auto-emulator.enabled` | To enable the auto configuration to start a local instance of the Datastore Emulator. | No | `false`

Author: Renaming `remulator-host` to just `host` may cause compatibilities issues. How to handle this? This exactly why I named the new emulator properties as `auto-emulator` haha

> | `spring.cloud.gcp.datastore.auto-emulator.enabled` | To enable the auto configuration to start a local instance of the Datastore Emulator. | No | `false`

Reviewer 2: What kind of incompatibilities? You mean backwards incompatibility? You can still leave `remulator-host` and mark it as deprecated.

> | `spring.cloud.gcp.datastore.auto-emulator.enabled` | To enable the auto configuration to start a local instance of the Datastore Emulator. | No | `false`

Author: But I will manually set the new configuration with the `emulator-host` if it is valid? We need to keep compatibility to those who already uses the `emulator-host`.

And will be weird since other modules like pubsub already uses `emulator-host`. Should we create an issue to fix them and continue with this new approach?

| `spring.cloud.gcp.datastore.auto-emulator.port` | The port to start the Datastore Emulator | No | `8081`

> | `spring.cloud.gcp.datastore.auto-emulator.consistency` | The https://cloud.google.com/sdk/gcloud/reference/beta/emulators/datastore/start?#--consistency[consistency] to use for the Datastore Emulator instance | No | `0.9`

Author: I can't.. The emulator starter lib set those properties hardcoded:
```java
   List<String> gcloudCommand = new ArrayList<>(Arrays.asList(GCLOUD_CMD_TEXT.split(" ")));
    gcloudCommand.add(GCLOUD_CMD_PORT_FLAG + "localhost:" + getPort());
    gcloudCommand.add(CONSISTENCY_FLAG + consistency);
    gcloudCommand.add("--no-store-on-disk");
    GcloudEmulatorRunner gcloudRunner =
        new GcloudEmulatorRunner(gcloudCommand, VERSION_PREFIX, MIN_VERSION);
    List<String> binCommand = new ArrayList<>(Arrays.asList(binName, "start"));
    binCommand.add("--testing");
    binCommand.add(BIN_CMD_PORT_FLAG + getPort());
    binCommand.add(CONSISTENCY_FLAG + consistency);
    if (gcdPath != null) {
      gcloudCommand.add("--data-dir=" + gcdPath.toString());
    }
```

We can change this later, but I think people will use this class mainly for integration tests purposes so the main objective here is to enable that first.

- an instance of all user defined repositories extending `CrudRepository`, `PagingAndSortingRepository`, and `DatastoreRepository` (an extension of `PagingAndSortingRepository` with additional Cloud Datastore features) when repositories are enabled
- an instance of `Datastore` from the Google Cloud Java Client for Datastore, for convenience and lower level API access

==== Datastore Emulator Autoconfiguration

> This Spring Boot autoconfiguration can also configure and start a local Datasstore Emulator server if enabled by property.

Reviewer 3: Typo: 
```suggestion
This Spring Boot autoconfiguration can also configure and start a local Datastore Emulator server if enabled by property.
```

> This is useful for integration test purposes and its usage is not recommended on production environments. The emulator will be properly stopped after the context shutdown.

Reviewer 3: One sentence per line, please: https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

When enabled the Datastore autoconfiguration itself will be forced to connect to the emulator instance if the `spring.cloud.gcp.datastore.host` property is empty.

> It will create an instance of `LocalDatastoreHelper` as bean that stores the `DatastoreOptions` to get the `Datastore` client connection to the emulator for convenience and lower level API for local access.

Reviewer 3: as **a** bean

- an instance of all user defined repositories extending `CrudRepository`, `PagingAndSortingRepository`, and `DatastoreRepository` (an extension of `PagingAndSortingRepository` with additional Cloud Datastore features) when repositories are enabled
- an instance of `Datastore` from the Google Cloud Java Client for Datastore, for convenience and lower level API access

==== Datastore Emulator Autoconfiguration

This Spring Boot autoconfiguration can also configure and start a local Datastore Emulator server if enabled by property.

> It is useful for integration testing and its usage is not recommended on production environments.

Reviewer 2: ```suggestion
It is useful for integration testing, but not for production.
```

Author: OK

> When enabled and if the `spring.cloud.gcp.datastore.host` property is empty, the Datastore autoconfiguration itself will be forced to connect to the local emulator instance.

Reviewer 2: ```suggestion
When enabled, the `spring.cloud.gcp.datastore.host` property will be ignored, and the Datastore autoconfiguration itself will be forced to connect to the autoconfigured local emulator instance.
```

Author: OK
